src/entities/sorted_state_adapter.ts

```diff
@@ -0,0 +1,167 @@
+import {
+  EntityState,
+  IdSelector,
+  Comparer,
+  EntityStateAdapter,
+  Update,
+  EntityMap
+} from './models'
+import { createStateOperator } from './state_adapter'
+import { createUnsortedStateAdapter } from './unsorted_state_adapter'
+import { selectIdValue } from './utils'
+
+export function createSortedStateAdapter<T>(
+  selectId: IdSelector<T>,
+  sort: Comparer<T>
+): EntityStateAdapter<T>
+export function createSortedStateAdapter<T>(selectId: any, sort: any): any {
+  type R = EntityState<T>
+
+  const { removeOne, removeMany, removeAll } = createUnsortedStateAdapter(
+    selectId
+  )
+
+  function addOneMutably(entity: T, state: R): void
+  function addOneMutably(entity: any, state: any): void {
+    return addManyMutably([entity], state)
+  }
+
+  function addManyMutably(newModels: T[], state: R): void
+  function addManyMutably(newModels: any[], state: any): void {
+    const models = newModels.filter(
+      model => !(selectIdValue(model, selectId) in state.entities)
+    )
+
+    if (models.length !== 0) {
+      merge(models, state)
+    }
+  }
+
+  function setAllMutably(models: T[], state: R): void
+  function setAllMutably(models: any[], state: any): void {
+    state.entities = {}
+    state.ids = []
+
+    addManyMutably(models, state)
+  }
+
+  function updateOneMutably(update: Update<T>, state: R): void
+  function updateOneMutably(update: any, state: any): void {
+    return updateManyMutably([update], state)
+  }
+
+  function takeUpdatedModel(models: T[], update: Update<T>, state: R): boolean
+  function takeUpdatedModel(models: any[], update: any, state: any): boolean {
+    if (!(update.id in state.entities)) {
+      return false
+    }
+
+    const original = state.entities[update.id]
+    const updated = Object.assign({}, original, update.changes)
+    const newKey = selectIdValue(updated, selectId)
+
+    delete state.entities[update.id]
+
+    models.push(updated)
+
+    return newKey !== update.id
+  }
+
+  function updateManyMutably(updates: Update<T>[], state: R): void
+  function updateManyMutably(updates: any[], state: any): void {
+    const models: T[] = []
+
+    updates.forEach(update => takeUpdatedModel(models, update, state))
+
+    if (models.length !== 0) {
+      merge(models, state)
+    }
+  }
+
+  function mapMutably(map: EntityMap<T>, state: R): void
+  function mapMutably(updatesOrMap: any, state: any): void {
+    const updates: Update<T>[] = state.ids.reduce(
+      (changes: any[], id: string | number) => {
+        const change = updatesOrMap(state.entities[id])
+        if (change !== state.entities[id]) {
```

Comment:
Also something that should be looked at, yeah.

Reply:
If there was some mutation inside, it should have been caught by Immer already, because we're either in a `produce()` _outside_ the entire entity logic (`createSlice`), or in a `produce()` up one level from here.  So, in the end Immer should have tracked that update regardless.